[Feature] Add docs about managed memory in flink.md (#1266)
apache/incubator-paimon: docs/content/engines/flink.md

* `MULTISET` is not supported.
* `MAP` is not supported as primary keys.

## Use Flink Managed Memory

> In order to improve the buffer pool for writers in Paimon tasks.Paimon tasks can create memory pools based on Executor Memory which will be managed by Executor, such as Managed Memory in Flink TaskManager. It will improve the stability and performance of sinks by managing writer buffers for multiple tasks through Executor.

Reviewer: ```suggestion
Paimon tasks can create memory pools based on executor memory which will be managed by Flink executor, such as managed memory in Flink task manager. It will improve the stability and performance of sinks by managing writer buffers for multiple tasks through executor.
```

> The following properties can be set if using the flink managed memory:

Reviewer: ```suggestion
The following properties can be set if using Flink managed memory:
```

| Option    | Default | Description                                                                                                                                                                    |
|------------|----------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | sink.use-managed-memory-allocator | false | If true, flink sink will use managed memory for merge tree; otherwise, it will create an independent memory allocator.                   |

Reviewer: ```suggestion
| sink.use-managed-memory-allocator | false | If true, Flink sink will use managed memory for merge tree; otherwise, it will create an independent memory allocator.                   |
```

Also, what does "independent memory allocator" mean? An independent memory allocator for each parallelism? You might also want to stress that this allocator will use heap memory, not Flink managed memory.

Author: Thank you for the review. I have made some modifications to enrich some of the content

> | sink.managed.writer-buffer-memory | 256M  | Weight of writer buffer in managed memory, Flink will compute the memory size, for writer according to the weight, the actual memory used depends on the running environment. |

Reviewer: "Weight of writer buffer"? This is not the real memory size? Could you provide a link to Flink's document about how managed memory are allocated to different operators?

Author: Thank you for the review. I have made some modifications

SELECT * FROM ....;
```

<span id="flinkallocator">**Use Allocator In Flink**</span>

The process of using Managed Memory Allocator in Flink is divided into two parts: compilation and runtime.

In compilation, Paimon creates sink operator in `FlinkSink`. Paimon can get `Transformation` from the sink operator, declare managed memory by `Transformation.declareManagedMemoryUseCaseAtOperatorScope(OPERATOR, weight of buffer)` according to options `sink.use-menaged-memory-allocator` and `sink.managed.writer-buffer-memory` . We can find a description of this method (`Transformation.declareManagedMemoryUseCaseAtOperatorScope(OPERATOR, weight of buffer)` ) from the [Javadoc](https://nightlies.apache.org/flink/flink-docs-release-1.17/api/java/org/apache/flink/api/dag/Transformation.html#declareManagedMemoryUseCaseAtOperatorScope-org.apache.flink.core.memory.ManagedMemoryUseCase-int-) ;

**note:** It should be noted that when the first parameter of the method is [`ManagedMemoryUseCase.OPERATOR`](https://nightlies.apache.org/flink/flink-docs-release-1.17/api/java/org/apache/flink/core/memory/ManagedMemoryUseCase.html#OPERATOR), the second parameter are defined as mebibyte values.

> In runtime, Paimon sink operator calculates writer buffer size based on `StreamConfig` and `MemoryManager` in `AbstractStoreWriteOperator.setup` , then creates `FlinkMemorySegmentAllocator` and `MemorySegmentPool` . Finally Paimon writer writes data to the memory pool and flush data to file system.

Reviewer: I can't quite get what you're trying to explain. As a user I have no interest in the implementation details. My main concern is that: is the memory size defined in `sink.managed.writer-buffer-memory` equal to the exact memory allocated to the write buffer in runtime? If not, how can I adjust this value to get my desired buffer size?

Author: The memory size defined in sink.managed.writer-buffer-memory equal to the exact memory allocated to write buffer in runtime, thanks
